### PR TITLE
feat(pl debug): pl debugging adapt odp directional routing

### PR DESCRIPTION
**Pull request by zijiacj**
#### What type of PR is this?
type-feature

#### What this PR does / why we need it:
 in pl debug, use odp's pointing route to ensure that the sql in debugeeSession and debugerSession are sent to the same observer

#### Which issue(s) this PR fixes:

Fixes #

#### Additional documentation e.g., usage docs, etc.:

```docs

```

### CODE REVIEW

File: server/odc-service/src/main/java/com/oceanbase/odc/service/pldebug/operator/DBPLOperators.java [hunk 1]
```diff
@@ -15,14 +15,22 @@
...
 import com.oceanbase.odc.core.shared.constant.ConnectType;
 import com.oceanbase.odc.core.shared.exception.UnsupportedException;
 
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
 public class DBPLOperators {
 
+    public static String odpSpecifiedRoutineEnabledVersionNumber = "3.1.11";
```
[the remark this answers is not in the document]
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/pldebug/operator/DBPLOperators.java:32]: ok

File: server/odc-service/src/main/java/com/oceanbase/odc/service/pldebug/operator/DBPLOperators.java [hunk 2]
```diff
@@ -33,8 +41,34 @@ public static DBPLOperator create(ConnectionSession connectionSession) {
...
     }
+
+    /**
+     * Get the OBProxy version number. If an exception occurs or the version does not support, return
+     * null.
+     *
+     * @param connectionSession
+     * @return
+     */
+    public static String getObProxyVersion(@NonNull ConnectionSession connectionSession) {
```
[the remark this answers is not in the document]
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/pldebug/operator/DBPLOperators.java:65]: ok

File: server/odc-service/src/main/java/com/oceanbase/odc/service/pldebug/operator/DBPLOperators.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/pldebug/operator/DBPLOperators.java:32]: ok

File: server/odc-service/src/main/java/com/oceanbase/odc/service/pldebug/session/AbstractDebugSession.java [hunk 3]
```diff
@@ -113,19 +119,40 @@ protected void acquireNewConnection(ConnectionSession connectionSession,
...
-        Integer port;
-        if (StringUtils.isBlank(config.getClusterName())) {
-            host = config.getHost();
-            port = config.getPort();
+        String host = config.getHost();
+        Integer port = config.getPort();
+        String url;
+
+        if (StringUtils.isBlank(config.getClusterName())
+                && connectionSession.getConnectType() != ConnectType.CLOUD_OB_ORACLE) {
```
[the remark this answers is not in the document]
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/pldebug/session/AbstractDebugSession.java:128]: The ob cloud oracle connection string also does not contain the cluster name

[the remark this answers is not in the document]
Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/pldebug/session/AbstractDebugSession.java:128]: I am still confused as to why you add this line...

[the remark this answers is not in the document]
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/pldebug/session/AbstractDebugSession.java:128]: If not, the ob cloud oracle connection is recognized as a direct observer connection

[the remark this answers is not in the document]
Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/pldebug/session/AbstractDebugSession.java:128]: 1. In the community ODC, there will not have cluster/tenant when you create a obcloud data source.
2. In the Cloud ODC, there will have a cluster/tenant.
Which case are you talking about?
Anyway, even if you are right, you should not use `&&` but `||`?

[the remark this answers is not in the document]
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/pldebug/session/AbstractDebugSession.java:128]: If the connection string is directly connected to a private cloud observer, the `clustername` must be null. Similarly, if the connection string is belong to public cloud, the clustername is also null. However, the public cloud connection must have obproxy, so the public cloud scenario should be excluded

[the remark this answers is not in the document]
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/pldebug/session/AbstractDebugSession.java:128]: obclient -h xxx -P xxx -uxxx@xxx#xxx

File: server/odc-service/src/main/java/com/oceanbase/odc/service/connection/util/ConnectionInfoUtil.java [hunk 4]
```diff
@@ -93,6 +94,22 @@ public static void initSessionVersion(@NonNull ConnectionSession connectionSessi
         log.debug("Init DB version completed.");
     }
 
+    public static void initOdpVersionIfExists(@NonNull ConnectionSession connectionSession) {
+        ConnectType connectType = connectionSession.getConnectType();
+        if (connectType == ConnectType.OB_MYSQL || connectType == ConnectType.OB_ORACLE
```
Comment (MarkPotato777) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/connection/util/ConnectionInfoUtil.java:99]: use `DialectType#isOceanbase` instead

File: server/odc-service/src/main/java/com/oceanbase/odc/service/pldebug/session/AbstractDebugSession.java [hunk 5]
```diff
@@ -113,19 +117,50 @@ protected void acquireNewConnection(ConnectionSession connectionSession,
...
-        if (StringUtils.isBlank(config.getClusterName())) {
-            host = config.getHost();
-            port = config.getPort();
-        } else {
+        String host = config.getHost();
+        Integer port = config.getPort();
+        String url;
+        if (connectionSession.getConnectType() == ConnectType.OB_ORACLE) {
+            if (StringUtils.isBlank(config.getClusterName())
+                    && connectionSession.getConnectType() != ConnectType.CLOUD_OB_ORACLE) {
```
Comment (MarkPotato777) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/pldebug/session/AbstractDebugSession.java:126]: You could delete this line `&& connectionSession.getConnectType() != ConnectType.CLOUD_OB_ORACLE `

File: server/odc-service/src/main/java/com/oceanbase/odc/service/pldebug/session/AbstractDebugSession.java [hunk 6]
```diff
@@ -113,19 +117,50 @@ protected void acquireNewConnection(ConnectionSession connectionSession,
...
+                }
+                // use direct connection observer
+                url = String.format("jdbc:%s://%s:%d/\"%s\"", OB_JDBC_PROTOCOL, host, port, schema);
+                return buildDataSource(config, initSqls, null, url);
+            }
+        } else if (connectionSession.getConnectType() == ConnectType.CLOUD_OB_ORACLE) {
             String directServerIp = getDirectServerIp(connectionSession);
-            host = directServerIp.split(":")[0];
-            port = Integer.parseInt(directServerIp.split(":")[1]);
+            String[] ipParts = directServerIp.split(":");
```
Comment (MarkPotato777) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/pldebug/session/AbstractDebugSession.java:150]: Duplicated code with the above.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/pldebug/session/AbstractDebugSession.java [hunk 7]
```diff
@@ -113,19 +117,50 @@ protected void acquireNewConnection(ConnectionSession connectionSession,
...
+            host = ipParts[0];
+            port = Integer.parseInt(ipParts[1]);
+            // use the specified routing function of odp
+            this.plDebugODPSpecifiedRoute = new PLDebugODPSpecifiedRoute(host, port);
```
Comment (MarkPotato777) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/pldebug/session/AbstractDebugSession.java:154]: Why no need to judge the odp version here?

File: server/integration-test/src/test/java/com/oceanbase/odc/service/pldebug/PLDebugTest.java [hunk 8]
```diff
@@ -107,7 +107,8 @@ public void test_call_procedure() throws Exception {
         params.add(paramC);
         DBProcedure odcProcedure = DBProcedure.of(null, "PROC", params);
 
-        CallProcedureCallBack procedureCallBack = new CallProcedureCallBack(odcProcedure, 30, new OracleSqlBuilder());
+        CallProcedureCallBack procedureCallBack =
+                new CallProcedureCallBack(odcProcedure, 30, new OracleSqlBuilder(), null);
```
Comment (MarkPotato777) [inline server/integration-test/src/test/java/com/oceanbase/odc/service/pldebug/PLDebugTest.java:111]: There is too much code with constructor parameters set to `null`. Try creating a new constructor that does not need pass `PLDebugODPSpecifiedRoute` and set it to false by default.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/connection/util/ConnectionInfoUtil.java (hunk 4 above)
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/connection/util/ConnectionInfoUtil.java:99]: ok

File: server/odc-service/src/main/java/com/oceanbase/odc/service/pldebug/session/AbstractDebugSession.java (hunk 5 above)
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/pldebug/session/AbstractDebugSession.java:126]: ok

File: server/odc-service/src/main/java/com/oceanbase/odc/service/pldebug/session/AbstractDebugSession.java (hunk 7 above)
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/pldebug/session/AbstractDebugSession.java:154]: cloud ob oracle's data source is previously verified in DBPLOperators#create, so it is not needed here.
image

File: server/integration-test/src/test/java/com/oceanbase/odc/service/pldebug/PLDebugTest.java (hunk 8 above)
Comment (zijiacj) [reply server/integration-test/src/test/java/com/oceanbase/odc/service/pldebug/PLDebugTest.java:111]: image
like that？

File: server/odc-service/src/main/java/com/oceanbase/odc/service/pldebug/session/AbstractDebugSession.java (hunk 6 above)
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/pldebug/session/AbstractDebugSession.java:150]: will be extracted into a method

File: server/odc-service/src/main/java/com/oceanbase/odc/service/pldebug/session/AbstractDebugSession.java (hunk 7 above)
Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/pldebug/session/AbstractDebugSession.java:166]: It's too tricky and hard to maintain. Put the version check logic together.

File: server/plugins/connect-plugin-api/src/main/java/com/oceanbase/odc/plugin/connect/api/InformationExtensionPoint.java [hunk 9]
```diff
@@ -28,4 +28,6 @@ public interface InformationExtensionPoint extends ExtensionPoint {
 
     String getDBVersion(Connection connection);
 
+    String getODPVersion(Connection connection);
```
Comment (MarkPotato777) [inline server/plugins/connect-plugin-api/src/main/java/com/oceanbase/odc/plugin/connect/api/InformationExtensionPoint.java:31]: ODP is a concept specific to OB and is not suitable to be included in a general interface.

Comment (zijiacj) [reply server/plugins/connect-plugin-api/src/main/java/com/oceanbase/odc/plugin/connect/api/InformationExtensionPoint.java:31]: image
As above, put method  `getODPVersion` in OBUtils

File: server/plugins/connect-plugin-oracle/src/main/java/com/oceanbase/odc/plugin/connect/oracle/OracleInformationExtension.java [hunk 10]
```diff
@@ -65,4 +65,5 @@ public String getDBVersion(Connection connection) {
             return "11.2.0.1.0";
         }
     }
+
```
Comment (MarkPotato777) [inline server/plugins/connect-plugin-oracle/src/main/java/com/oceanbase/odc/plugin/connect/oracle/OracleInformationExtension.java:68]: Avoid empty line updates. Same as below.

Comment (zijiacj) [reply server/plugins/connect-plugin-oracle/src/main/java/com/oceanbase/odc/plugin/connect/oracle/OracleInformationExtension.java]: done

**Review by MarkPotato777 — APPROVED** [review]
LGTM

**Review by LuckyPickleZZ — APPROVED** [review]

**Review by PeachThinking — APPROVED** [review]
lgtm

**Review by yhilmare — APPROVED** [review]
lgtm

**Pull request merged**